// adaptivecards/lib/activity-request.js
export var ActivityRequestTrigger;
(function (ActivityRequestTrigger) {
    ActivityRequestTrigger["Automatic"] = "automatic";
    ActivityRequestTrigger["Manual"] = "manual";
})(ActivityRequestTrigger || (ActivityRequestTrigger = {}));
export class ActivityRequestError {
    constructor(code, message) {
        this.code = code;
        this.message = message;
    }
}
export class ActivityResponse {
    constructor(request) {
        this.request = request;
    }
}
export class SuccessResponse extends ActivityResponse {
    constructor(request, rawContent) {
        super(request);
        this.request = request;
        this.rawContent = rawContent;
    }
}
export class ErrorResponse extends ActivityResponse {
    constructor(request, error) {
        super(request);
        this.request = request;
        this.error = error;
    }
}
export class LoginRequestResponse extends ActivityResponse {
    constructor(request, auth) {
        super(request);
        this.request = request;
        this._auth = auth;
        for (const button of this._auth.buttons) {
            if (button.type === "signin" && button.value !== undefined) {
                try {
                    new URL(button.value);
                    this.signinButton = button;
                    break;
                }
                catch (e) {
                    // Ignore parsing error
                }
            }
        }
    }
    get tokenExchangeResource() {
        return this._auth.tokenExchangeResource;
    }
}
