// adaptivecards/lib/adaptive-applet.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as Enums from "./enums";
import * as Utils from "./utils";
import { GlobalSettings } from "./shared";
import { ActivityRequestTrigger, SuccessResponse, ErrorResponse, LoginRequestResponse } from "./activity-request";
import { Strings } from "./strings";
import { SubmitAction, ExecuteAction, SerializationContext, AdaptiveCard, UniversalAction, ChoiceSetInput, DataQuery } from "./card-elements";
import { Versions } from "./serialization";
function logEvent(level, message, ...optionalParams) {
    if (GlobalSettings.applets.logEnabled) {
        if (GlobalSettings.applets.onLogEvent) {
            GlobalSettings.applets.onLogEvent(level, message, optionalParams);
        }
        else {
            /* eslint-disable no-console */
            switch (level) {
                case Enums.LogLevel.Warning:
                    console.warn(message, optionalParams);
                    break;
                case Enums.LogLevel.Error:
                    console.error(message, optionalParams);
                    break;
                default:
                    console.log(message, optionalParams);
                    break;
            }
            /* eslint-enable no-console */
        }
    }
}
class ActivityRequest {
    constructor(action, trigger, consecutiveActions) {
        this.action = action;
        this.trigger = trigger;
        this.consecutiveActions = consecutiveActions;
        this.attemptNumber = 0;
    }
    // eslint-disable-next-line @typescript-eslint/require-await
    retryAsync() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.onSend) {
                this.onSend(this);
            }
        });
    }
}
export class AdaptiveApplet {
    displayCard(card) {
        if (card.renderedElement) {
            Utils.clearElementChildren(this._cardHostElement);
            this._refreshButtonHostElement.style.display = "none";
            this._cardHostElement.appendChild(card.renderedElement);
        }
        else {
            throw new Error("displayCard: undefined card.");
        }
    }
    showManualRefreshButton(refreshAction) {
        const displayBuiltInManualRefreshButton = this.onShowManualRefreshButton
            ? this.onShowManualRefreshButton(this)
            : true;
        if (displayBuiltInManualRefreshButton) {
            this._refreshButtonHostElement.style.display = "none";
            let renderedRefreshButton = undefined;
            if (this.onRenderManualRefreshButton) {
                renderedRefreshButton = this.onRenderManualRefreshButton(this);
            }
            else {
                let message = Strings.runtime.refreshThisCard();
                if (GlobalSettings.applets.refresh.mode === Enums.RefreshMode.Automatic) {
                    let autoRefreshPausedMessage = Strings.runtime.automaticRefreshPaused();
                    if (autoRefreshPausedMessage[autoRefreshPausedMessage.length - 1] !== " ") {
                        autoRefreshPausedMessage += " ";
                    }
                    message = Strings.runtime.clckToRestartAutomaticRefresh();
                }
                const cardPayload = {
                    type: "AdaptiveCard",
                    version: "1.2",
                    body: [
                        {
                            type: "RichTextBlock",
                            horizontalAlignment: "right",
                            inlines: [
                                {
                                    type: "TextRun",
                                    text: message,
                                    selectAction: {
                                        type: "Action.Submit",
                                        id: "refreshCard"
                                    }
                                }
                            ]
                        }
                    ]
                };
                const card = new AdaptiveCard();
                card.parse(cardPayload, new SerializationContext(Versions.v1_2));
                card.onExecuteAction = (action) => {
                    if (action.id === "refreshCard") {
                        Utils.clearElementChildren(this._refreshButtonHostElement);
                        this.internalExecuteAction(refreshAction, ActivityRequestTrigger.Automatic, 0);
                    }
                };
                renderedRefreshButton = card.render();
            }
            if (renderedRefreshButton) {
                Utils.clearElementChildren(this._refreshButtonHostElement);
                this._refreshButtonHostElement.appendChild(renderedRefreshButton);
                this._refreshButtonHostElement.style.removeProperty("display");
            }
        }
    }
    createActivityRequest(action, trigger, consecutiveActions) {
        if (this.card) {
            const request = new ActivityRequest(action, trigger, consecutiveActions);
            request.onSend = (sender) => {
                sender.attemptNumber++;
                void this.internalSendActivityRequestAsync(request);
            };
            const cancel = this.onPrepareActivityRequest
                ? !this.onPrepareActivityRequest(this, request, action)
                : false;
            return cancel ? undefined : request;
        }
        else {
            throw new Error("createActivityRequest: no card has been set.");
        }
    }
    createMagicCodeInputCard(attemptNumber) {
        const payload = {
            type: "AdaptiveCard",
            version: "1.0",
            body: [
                {
                    type: "TextBlock",
                    color: "attention",
                    text: attemptNumber === 1 ? undefined : "That didn't work... let's try again.",
                    wrap: true,
                    horizontalAlignment: "center"
                },
                {
                    type: "TextBlock",
                    text: 'Please login in the popup. You will obtain a magic code. Paste that code below and select "Submit"',
                    wrap: true,
                    horizontalAlignment: "center"
                },
                {
                    type: "Input.Text",
                    id: "magicCode",
                    placeholder: "Enter magic code"
                },
                {
                    type: "ActionSet",
                    horizontalAlignment: "center",
                    actions: [
                        {
                            type: "Action.Submit",
                            id: AdaptiveApplet._submitMagicCodeActionId,
                            title: "Submit"
                        },
                        {
                            type: "Action.Submit",
                            id: AdaptiveApplet._cancelMagicCodeAuthActionId,
                            title: "Cancel"
                        }
                    ]
                }
            ]
        };
        const card = new AdaptiveCard();
        card.parse(payload);
        return card;
    }
    cancelAutomaticRefresh() {
        if (this._allowAutomaticCardUpdate) {
            logEvent(Enums.LogLevel.Warning, "Automatic card refresh has been cancelled as a result of the user interacting with the card.");
        }
        this._allowAutomaticCardUpdate = false;
    }
    createSerializationContext() {
        return this.onCreateSerializationContext
            ? this.onCreateSerializationContext(this)
            : new SerializationContext();
    }
    internalSetCard(payload, consecutiveActions) {
        if (typeof payload === "object" && payload["type"] === "AdaptiveCard") {
            this._cardPayload = payload;
        }
        if (this._cardPayload) {
            try {
                const card = new AdaptiveCard();
                if (this.hostConfig) {
                    card.hostConfig = this.hostConfig;
                }
                const serializationContext = this.createSerializationContext();
                card.parse(this._cardPayload, serializationContext);
                const doChangeCard = this.onCardChanging
                    ? this.onCardChanging(this, this._cardPayload)
                    : true;
                if (doChangeCard) {
                    this._card = card;
                    if (this._card.authentication &&
                        this._card.authentication.tokenExchangeResource &&
                        this.onPrefetchSSOToken) {
                        this.onPrefetchSSOToken(this, this._card.authentication.tokenExchangeResource);
                    }
                    this._card.onExecuteAction = (action) => {
                        // If the user takes an action, cancel any pending automatic refresh
                        this.cancelAutomaticRefresh();
                        this.internalExecuteAction(action, ActivityRequestTrigger.Manual, 0);
                    };
                    this._card.onInputValueChanged = (input) => {
                        var _a, _b, _c;
                        // If the user modifies an input, cancel any pending automatic refresh
                        this.cancelAutomaticRefresh();
                        if (input instanceof ChoiceSetInput && input.isDynamicTypeahead()) {
                            const filter = input.getFilterForDynamicSearch();
                            if (filter) {
                                const dataQueryAction = new DataQuery();
                                dataQueryAction.filter = filter;
                                dataQueryAction.dataset = ((_a = input.choicesData) === null || _a === void 0 ? void 0 : _a.dataset) || "";
                                dataQueryAction.count = (_b = input.choicesData) === null || _b === void 0 ? void 0 : _b.count;
                                dataQueryAction.skip = (_c = input.choicesData) === null || _c === void 0 ? void 0 : _c.skip;
                                this._choiceSet = input;
                                this.internalExecuteAction(dataQueryAction, ActivityRequestTrigger.Manual, 0 // consecutiveActions
                                );
                            }
                        }
                    };
                    this._card.render();
                    if (this._card.renderedElement) {
                        this.displayCard(this._card);
                        if (this.onCardChanged) {
                            this.onCardChanged(this);
                        }
                        if (this._card.refresh) {
                            if (GlobalSettings.applets.refresh.mode ===
                                Enums.RefreshMode.Automatic &&
                                consecutiveActions <
                                    GlobalSettings.applets.refresh
                                        .maximumConsecutiveAutomaticRefreshes) {
                                if (GlobalSettings.applets.refresh.timeBetweenAutomaticRefreshes <=
                                    0) {
                                    logEvent(Enums.LogLevel.Info, "Triggering automatic card refresh number " +
                                        (consecutiveActions + 1));
                                    this.internalExecuteAction(this._card.refresh.action, ActivityRequestTrigger.Automatic, consecutiveActions + 1);
                                }
                                else {
                                    logEvent(Enums.LogLevel.Info, "Scheduling automatic card refresh number " +
                                        (consecutiveActions + 1) +
                                        " in " +
                                        GlobalSettings.applets.refresh
                                            .timeBetweenAutomaticRefreshes +
                                        "ms");
                                    const action = this._card.refresh.action;
                                    this._allowAutomaticCardUpdate = true;
                                    window.setTimeout(() => {
                                        if (this._allowAutomaticCardUpdate) {
                                            this.internalExecuteAction(action, ActivityRequestTrigger.Automatic, consecutiveActions + 1);
                                        }
                                    }, GlobalSettings.applets.refresh.timeBetweenAutomaticRefreshes);
                                }
                            }
                            else if (GlobalSettings.applets.refresh.mode !== Enums.RefreshMode.Disabled) {
                                if (consecutiveActions > 0) {
                                    logEvent(Enums.LogLevel.Warning, "Stopping automatic refreshes after " +
                                        consecutiveActions +
                                        " consecutive refreshes.");
                                }
                                else {
                                    logEvent(Enums.LogLevel.Warning, "The card has a refresh section, but automatic refreshes are disabled.");
                                }
                                if (GlobalSettings.applets.refresh
                                    .allowManualRefreshesAfterAutomaticRefreshes ||
                                    GlobalSettings.applets.refresh.mode === Enums.RefreshMode.Manual) {
                                    logEvent(Enums.LogLevel.Info, "Showing manual refresh button.");
                                    this.showManualRefreshButton(this._card.refresh.action);
                                }
                            }
                        }
                    }
                }
            }
            catch (error) {
                // Ignore all errors
                logEvent(Enums.LogLevel.Error, "setCard: " + error);
            }
        }
    }
    internalExecuteAction(action, trigger, consecutiveActions) {
        if (action instanceof UniversalAction) {
            if (this.channelAdapter) {
                const request = this.createActivityRequest(action, trigger, consecutiveActions);
                if (request) {
                    void request.retryAsync();
                }
            }
            else {
                throw new Error("internalExecuteAction: No channel adapter set.");
            }
        }
        if (this.onAction) {
            this.onAction(this, action);
        }
    }
    createProgressOverlay(request) {
        if (!this._progressOverlay) {
            if (this.onCreateProgressOverlay) {
                this._progressOverlay = this.onCreateProgressOverlay(this, request);
            }
            else {
                this._progressOverlay = document.createElement("div");
                this._progressOverlay.className = "aaf-progress-overlay";
                const spinner = document.createElement("div");
                spinner.className = "aaf-spinner";
                spinner.style.width = "28px";
                spinner.style.height = "28px";
                this._progressOverlay.appendChild(spinner);
            }
        }
        return this._progressOverlay;
    }
    removeProgressOverlay(request) {
        if (this.onRemoveProgressOverlay) {
            this.onRemoveProgressOverlay(this, request);
        }
        if (this._progressOverlay !== undefined) {
            this.renderedElement.removeChild(this._progressOverlay);
            this._progressOverlay = undefined;
        }
    }
    activityRequestSucceeded(response, parsedContent) {
        if (this.onActivityRequestSucceeded) {
            this.onActivityRequestSucceeded(this, response, parsedContent);
        }
    }
    activityRequestFailed(response) {
        return this.onActivityRequestFailed
            ? this.onActivityRequestFailed(this, response)
            : GlobalSettings.applets.defaultTimeBetweenRetryAttempts;
    }
    showAuthCodeInputDialog(request) {
        const showBuiltInAuthCodeInputCard = this.onShowAuthCodeInputDialog
            ? this.onShowAuthCodeInputDialog(this, request)
            : true;
        if (showBuiltInAuthCodeInputCard) {
            const authCodeInputCard = this.createMagicCodeInputCard(request.attemptNumber);
            authCodeInputCard.render();
            authCodeInputCard.onExecuteAction = (submitMagicCodeAction) => {
                if (this.card && submitMagicCodeAction instanceof SubmitAction) {
                    switch (submitMagicCodeAction.id) {
                        case AdaptiveApplet._submitMagicCodeActionId:
                            let authCode = undefined;
                            if (submitMagicCodeAction.data &&
                                typeof submitMagicCodeAction.data["magicCode"] === "string") {
                                authCode = submitMagicCodeAction.data["magicCode"];
                            }
                            if (authCode) {
                                this.displayCard(this.card);
                                request.authCode = authCode;
                                void request.retryAsync();
                            }
                            else {
                                alert("Please enter the magic code you received.");
                            }
                            break;
                        case AdaptiveApplet._cancelMagicCodeAuthActionId:
                            logEvent(Enums.LogLevel.Warning, "Authentication cancelled by user.");
                            this.displayCard(this.card);
                            break;
                        default:
                            logEvent(Enums.LogLevel.Error, "Unexpected action taken from magic code input card (id = " +
                                submitMagicCodeAction.id +
                                ")");
                            alert(Strings.magicCodeInputCard.somethingWentWrong());
                            break;
                    }
                }
            };
            this.displayCard(authCodeInputCard);
        }
    }
    internalSendActivityRequestAsync(request) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.channelAdapter) {
                throw new Error("internalSendActivityRequestAsync: channelAdapter is not set.");
            }
            const { action } = request;
            if (action instanceof ExecuteAction) {
                yield this.internalSendExecuteRequestAsync(request);
            }
            else if (action instanceof DataQuery) {
                yield this.internalSendDataQueryRequestAsync(request);
            }
            else {
                throw new Error("internalSendActivityRequestAsync: Unhandled Action Type");
            }
        });
    }
    internalSendExecuteRequestAsync(request) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.channelAdapter) {
                throw new Error("internalSendExecuteRequestAsync: channelAdapter is not set.");
            }
            const overlay = this.createProgressOverlay(request);
            if (overlay !== undefined) {
                this.renderedElement.appendChild(overlay);
            }
            let done = false;
            while (!done) {
                let response = undefined;
                if (request.attemptNumber === 1) {
                    logEvent(Enums.LogLevel.Info, "Sending activity request to channel (attempt " + request.attemptNumber + ")");
                }
                else {
                    logEvent(Enums.LogLevel.Info, "Re-sending activity request to channel (attempt " + request.attemptNumber + ")");
                }
                try {
                    response = yield this.channelAdapter.sendRequestAsync(request);
                }
                catch (error) {
                    logEvent(Enums.LogLevel.Error, "Activity request failed: " + error);
                    this.removeProgressOverlay(request);
                    done = true;
                }
                if (response) {
                    if (response instanceof SuccessResponse) {
                        this.removeProgressOverlay(request);
                        if (response.rawContent === undefined) {
                            throw new Error("internalSendExecuteRequestAsync: Action.Execute result is undefined");
                        }
                        let parsedContent = response.rawContent;
                        try {
                            parsedContent = JSON.parse(response.rawContent);
                        }
                        catch (_a) {
                            // Leave parseContent as is
                        }
                        if (typeof parsedContent === "string") {
                            logEvent(Enums.LogLevel.Info, "The activity request returned a string after " +
                                request.attemptNumber +
                                " attempt(s).");
                            this.activityRequestSucceeded(response, parsedContent);
                        }
                        else if (typeof parsedContent === "object" &&
                            parsedContent["type"] === "AdaptiveCard") {
                            logEvent(Enums.LogLevel.Info, "The activity request returned an Adaptive Card after " +
                                request.attemptNumber +
                                " attempt(s).");
                            this.internalSetCard(parsedContent, request.consecutiveActions);
                            this.activityRequestSucceeded(response, this.card);
                        }
                        else {
                            throw new Error("internalSendExecuteRequestAsync: Action.Execute result is of unsupported type (" +
                                typeof response.rawContent +
                                ")");
                        }
                        done = true;
                    }
                    else if (response instanceof ErrorResponse) {
                        const retryIn = this.activityRequestFailed(response);
                        if (retryIn >= 0 &&
                            request.attemptNumber < GlobalSettings.applets.maximumRetryAttempts) {
                            logEvent(Enums.LogLevel.Warning, `Activity request failed: ${response.error.message}. Retrying in ${retryIn}ms`);
                            request.attemptNumber++;
                            yield new Promise((resolve, _reject) => {
                                window.setTimeout(() => {
                                    resolve();
                                }, retryIn);
                            });
                        }
                        else {
                            logEvent(Enums.LogLevel.Error, `Activity request failed: ${response.error.message}. Giving up after ${request.attemptNumber} attempt(s)`);
                            this.removeProgressOverlay(request);
                            done = true;
                        }
                    }
                    else if (response instanceof LoginRequestResponse) {
                        logEvent(Enums.LogLevel.Info, "The activity request returned a LoginRequestResponse after " +
                            request.attemptNumber +
                            " attempt(s).");
                        if (request.attemptNumber <= GlobalSettings.applets.maximumRetryAttempts) {
                            let attemptOAuth = true;
                            if (response.tokenExchangeResource && this.onSSOTokenNeeded) {
                                // Attempt to use SSO. The host will return true if it can handle SSO, in which case
                                // we bypass OAuth
                                attemptOAuth = !this.onSSOTokenNeeded(this, request, response.tokenExchangeResource);
                            }
                            if (attemptOAuth) {
                                // Attempt to use OAuth
                                this.removeProgressOverlay(request);
                                if (response.signinButton === undefined) {
                                    throw new Error("internalSendExecuteRequestAsync: the login request doesn't contain a valid signin URL.");
                                }
                                logEvent(Enums.LogLevel.Info, "Login required at " + response.signinButton.value);
                                if (this.onShowSigninPrompt) {
                                    // Bypass the built-in auth prompt if the host app handles it
                                    this.onShowSigninPrompt(this, request, response.signinButton);
                                }
                                else {
                                    this.showAuthCodeInputDialog(request);
                                    const left = window.screenX +
                                        (window.outerWidth - GlobalSettings.applets.authPromptWidth) /
                                            2;
                                    const top = window.screenY +
                                        (window.outerHeight - GlobalSettings.applets.authPromptHeight) /
                                            2;
                                    window.open(response.signinButton.value, response.signinButton.title
                                        ? response.signinButton.title
                                        : "Sign in", `width=${GlobalSettings.applets.authPromptWidth},height=${GlobalSettings.applets.authPromptHeight},left=${left},top=${top}`);
                                }
                            }
                        }
                        else {
                            logEvent(Enums.LogLevel.Error, "Authentication failed. Giving up after " +
                                request.attemptNumber +
                                " attempt(s)");
                            alert(Strings.magicCodeInputCard.authenticationFailed());
                        }
                        // Exit the loop. After a LoginRequestResponse, the host app is responsible for retrying the request
                        break;
                    }
                    else {
                        throw new Error("Unhandled response type: " + JSON.stringify(response));
                    }
                }
            }
        });
    }
    internalSendDataQueryRequestAsync(request) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.channelAdapter) {
                throw new Error("internalSendDataQueryRequestAsync: channel adapter not set");
            }
            const filter = request.action.filter;
            if (this._choiceSet) {
                this._choiceSet.showLoadingIndicator();
                let response = undefined;
                try {
                    response = yield this.channelAdapter.sendRequestAsync(request);
                }
                catch (error) {
                    this._choiceSet.showErrorIndicator(filter, "Unable to load");
                    logEvent(Enums.LogLevel.Error, "Activity request failed: " + error);
                }
                if (response) {
                    if (response instanceof SuccessResponse) {
                        const rawResponse = response.rawContent;
                        if (rawResponse) {
                            let parsedResponse;
                            try {
                                parsedResponse = JSON.parse(rawResponse);
                            }
                            catch (error) {
                                this._choiceSet.showErrorIndicator(filter, "Unable to load");
                                throw new Error("internalSendDataQueryRequestAsync: Cannot parse response object: " +
                                    rawResponse);
                            }
                            if (typeof parsedResponse === "object") {
                                this._choiceSet.renderChoices(filter, parsedResponse);
                                this.activityRequestSucceeded(response, rawResponse);
                            }
                            else {
                                this._choiceSet.showErrorIndicator(filter, "Error loading results");
                                throw new Error("internalSendDataQueryRequestAsync: Data.Query result is of unsupported type (" +
                                    typeof rawResponse +
                                    ")");
                            }
                        }
                        else {
                            this._choiceSet.showErrorIndicator(filter, "Unable to load");
                            throw new Error("internalSendDataQueryRequestAsync: Data.Query result is undefined");
                        }
                    }
                    else if (response instanceof ErrorResponse) {
                        this._choiceSet.showErrorIndicator(filter, response.error.message || "Error loading results");
                        logEvent(Enums.LogLevel.Error, `Activity request failed: ${response.error.message}.`);
                        this.activityRequestFailed(response);
                    }
                    else {
                        this._choiceSet.showErrorIndicator(filter, "Unable to load");
                        throw new Error("Unhandled response type: " + JSON.stringify(response));
                    }
                }
                else {
                    this._choiceSet.removeLoadingIndicator();
                }
            }
        });
    }
    constructor() {
        this._allowAutomaticCardUpdate = false;
        this.renderedElement = document.createElement("div");
        this.renderedElement.className = "aaf-cardHost";
        this.renderedElement.style.position = "relative";
        this.renderedElement.style.display = "flex";
        this.renderedElement.style.flexDirection = "column";
        this._cardHostElement = document.createElement("div");
        this._refreshButtonHostElement = document.createElement("div");
        this._refreshButtonHostElement.className = "aaf-refreshButtonHost";
        this._refreshButtonHostElement.style.display = "none";
        this.renderedElement.appendChild(this._cardHostElement);
        this.renderedElement.appendChild(this._refreshButtonHostElement);
    }
    refreshCard() {
        if (this._card && this._card.refresh) {
            this.internalExecuteAction(this._card.refresh.action, ActivityRequestTrigger.Manual, 0);
        }
    }
    setCard(payload) {
        this.internalSetCard(payload, 0);
    }
    get card() {
        return this._card;
    }
}
AdaptiveApplet._submitMagicCodeActionId = "submitMagicCode";
AdaptiveApplet._cancelMagicCodeAuthActionId = "cancelMagicCodeAuth";
