// adaptivecards/lib/table.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { StylableCardElementContainer, Container, ContainerStyleProperty } from "./card-elements";
import { HorizontalAlignment, VerticalAlignment, SizeUnit, ValidationEvent, Spacing } from "./enums";
import { GlobalRegistry } from "./registry";
import { BoolProperty, CustomProperty, EnumProperty, property, SerializableObject, SerializableObjectCollectionProperty, Versions } from "./serialization";
import { SizeAndUnit, PaddingDefinition } from "./shared";
import { Strings } from "./strings";
import { stringToCssColor } from "./utils";
export class TableColumnDefinition extends SerializableObject {
    constructor() {
        //#region Schema
        super(...arguments);
        this.width = new SizeAndUnit(1, SizeUnit.Weight);
    }
    getSchemaKey() {
        return "ColumnDefinition";
    }
}
TableColumnDefinition.horizontalCellContentAlignmentProperty = new EnumProperty(Versions.v1_5, "horizontalCellContentAlignment", HorizontalAlignment);
TableColumnDefinition.verticalCellContentAlignmentProperty = new EnumProperty(Versions.v1_5, "verticalCellContentAlignment", VerticalAlignment);
TableColumnDefinition.widthProperty = new CustomProperty(Versions.v1_5, "width", (sender, prop, source, context) => {
    let result = prop.defaultValue;
    const value = source[prop.name];
    let invalidWidth = false;
    if (typeof value === "number" && !isNaN(value)) {
        result = new SizeAndUnit(value, SizeUnit.Weight);
    }
    else if (typeof value === "string") {
        try {
            result = SizeAndUnit.parse(value);
        }
        catch (e) {
            invalidWidth = true;
        }
    }
    else {
        invalidWidth = true;
    }
    if (invalidWidth) {
        context.logParseEvent(sender, ValidationEvent.InvalidPropertyValue, Strings.errors.invalidColumnWidth(value));
    }
    return result;
}, (sender, property, target, value, context) => {
    if (value.unit === SizeUnit.Pixel) {
        context.serializeValue(target, "width", value.physicalSize + "px");
    }
    else {
        context.serializeNumber(target, "width", value.physicalSize);
    }
}, new SizeAndUnit(1, SizeUnit.Weight));
__decorate([
    property(TableColumnDefinition.horizontalCellContentAlignmentProperty)
], TableColumnDefinition.prototype, "horizontalCellContentAlignment", void 0);
__decorate([
    property(TableColumnDefinition.verticalCellContentAlignmentProperty)
], TableColumnDefinition.prototype, "verticalCellContentAlignment", void 0);
__decorate([
    property(TableColumnDefinition.widthProperty)
], TableColumnDefinition.prototype, "width", void 0);
export class StylableContainer extends StylableCardElementContainer {
    constructor() {
        super(...arguments);
        this._items = [];
    }
    parseItem(source, context) {
        return context.parseCardObject(this, source, [], // Forbidden types not supported for elements for now
        !this.isDesignMode(), (typeName) => {
            return this.createItemInstance(typeName);
        }, (typeName, _errorType) => {
            context.logParseEvent(undefined, ValidationEvent.ElementTypeNotAllowed, Strings.errors.elementTypeNotAllowed(typeName));
        });
    }
    internalAddItem(item) {
        if (!item.parent) {
            this._items.push(item);
            item.setParent(this);
        }
        else {
            throw new Error(Strings.errors.elementAlreadyParented());
        }
    }
    internalRemoveItem(item) {
        const itemIndex = this._items.indexOf(item);
        if (itemIndex >= 0) {
            this._items.splice(itemIndex, 1);
            item.setParent(undefined);
            this.updateLayout();
            return true;
        }
        return false;
    }
    internalIndexOf(item) {
        return this._items.indexOf(item);
    }
    internalParse(source, context) {
        super.internalParse(source, context);
        this._items = [];
        const items = source[this.getCollectionPropertyName()];
        if (Array.isArray(items)) {
            for (const item of items) {
                const instance = this.parseItem(item, context);
                if (instance) {
                    this._items.push(instance);
                }
            }
        }
    }
    internalToJSON(target, context) {
        super.internalToJSON(target, context);
        context.serializeArray(target, this.getCollectionPropertyName(), this._items);
    }
    removeItem(item) {
        return this.internalRemoveItem(item);
    }
    getItemCount() {
        return this._items.length;
    }
    getItemAt(index) {
        return this._items[index];
    }
    getFirstVisibleRenderedItem() {
        return this.getItemCount() > 0 ? this.getItemAt(0) : undefined;
    }
    getLastVisibleRenderedItem() {
        return this.getItemCount() > 0 ? this.getItemAt(this.getItemCount() - 1) : undefined;
    }
}
export class TableCell extends Container {
    constructor() {
        super(...arguments);
        this._columnIndex = -1;
        this._cellType = "data";
    }
    getHasBorder() {
        return this.parentRow.parentTable.showGridLines;
    }
    applyBorder() {
        if (this.renderedElement && this.getHasBorder()) {
            const styleDefinition = this.hostConfig.containerStyles.getStyleByName(this.parentRow.parentTable.gridStyle);
            if (styleDefinition.borderColor) {
                const borderColor = stringToCssColor(styleDefinition.borderColor);
                if (borderColor) {
                    this.renderedElement.style.borderRight = "1px solid " + borderColor;
                    this.renderedElement.style.borderBottom = "1px solid " + borderColor;
                }
            }
        }
    }
    getDefaultPadding() {
        return this.getHasBackground() || this.getHasBorder()
            ? new PaddingDefinition(Spacing.Small, Spacing.Small, Spacing.Small, Spacing.Small)
            : super.getDefaultPadding();
    }
    internalRender() {
        const cellElement = super.internalRender();
        if (cellElement) {
            cellElement.setAttribute("role", this.cellType === "data" ? "cell" : "columnheader");
            cellElement.style.minWidth = "0";
            if (this.cellType === "header") {
                cellElement.setAttribute("scope", "col");
            }
        }
        return cellElement;
    }
    shouldSerialize(_context) {
        return true;
    }
    getJsonTypeName() {
        return "TableCell";
    }
    getEffectiveTextStyleDefinition() {
        if (this.cellType === "header") {
            return this.hostConfig.textStyles.columnHeader;
        }
        return super.getEffectiveTextStyleDefinition();
    }
    getEffectiveHorizontalAlignment() {
        if (this.horizontalAlignment !== undefined) {
            return this.horizontalAlignment;
        }
        if (this.parentRow.horizontalCellContentAlignment !== undefined) {
            return this.parentRow.horizontalCellContentAlignment;
        }
        if (this.columnIndex >= 0) {
            const horizontalAlignment = this.parentRow.parentTable.getColumnAt(this.columnIndex).horizontalCellContentAlignment;
            if (horizontalAlignment !== undefined) {
                return horizontalAlignment;
            }
        }
        if (this.parentRow.parentTable.horizontalCellContentAlignment !== undefined) {
            return this.parentRow.parentTable.horizontalCellContentAlignment;
        }
        return super.getEffectiveHorizontalAlignment();
    }
    getEffectiveVerticalContentAlignment() {
        if (this.verticalContentAlignment !== undefined) {
            return this.verticalContentAlignment;
        }
        if (this.parentRow.verticalCellContentAlignment !== undefined) {
            return this.parentRow.verticalCellContentAlignment;
        }
        if (this.columnIndex >= 0) {
            const verticalAlignment = this.parentRow.parentTable.getColumnAt(this.columnIndex).verticalCellContentAlignment;
            if (verticalAlignment !== undefined) {
                return verticalAlignment;
            }
        }
        if (this.parentRow.parentTable.verticalCellContentAlignment !== undefined) {
            return this.parentRow.parentTable.verticalCellContentAlignment;
        }
        return super.getEffectiveVerticalContentAlignment();
    }
    get columnIndex() {
        return this._columnIndex;
    }
    get cellType() {
        return this._cellType;
    }
    get parentRow() {
        return this.parent;
    }
    get isStandalone() {
        return false;
    }
}
export class TableRow extends StylableContainer {
    //#endregion
    getDefaultPadding() {
        return new PaddingDefinition(Spacing.None, Spacing.None, Spacing.None, Spacing.None);
    }
    applyBackground() {
        if (this.renderedElement) {
            const styleDefinition = this.hostConfig.containerStyles.getStyleByName(this.style, this.hostConfig.containerStyles.getStyleByName(this.defaultStyle));
            if (styleDefinition.backgroundColor) {
                const bgColor = stringToCssColor(styleDefinition.backgroundColor);
                if (bgColor) {
                    this.renderedElement.style.backgroundColor = bgColor;
                }
            }
        }
    }
    getCollectionPropertyName() {
        return "cells";
    }
    createItemInstance(typeName) {
        return !typeName || typeName === "TableCell" ? new TableCell() : undefined;
    }
    internalRender() {
        const isFirstRow = this.getIsFirstRow();
        const cellSpacing = this.hostConfig.table.cellSpacing;
        const rowElement = document.createElement("div");
        rowElement.setAttribute("role", "row");
        rowElement.style.display = "flex";
        rowElement.style.flexDirection = "row";
        for (let i = 0; i < Math.min(this.getItemCount(), this.parentTable.getColumnCount()); i++) {
            const cell = this.getItemAt(i);
            // Cheating a bit in order to keep cellType read-only
            cell["_columnIndex"] = i;
            cell["_cellType"] =
                this.parentTable.firstRowAsHeaders && isFirstRow ? "header" : "data";
            const renderedCell = cell.render();
            if (renderedCell) {
                const column = this.parentTable.getColumnAt(i);
                if (column.computedWidth.unit === SizeUnit.Pixel) {
                    renderedCell.style.flex = "0 0 " + column.computedWidth.physicalSize + "px";
                }
                else {
                    renderedCell.style.flex = "1 1 " + column.computedWidth.physicalSize + "%";
                }
                if (i > 0 && !this.parentTable.showGridLines && cellSpacing > 0) {
                    renderedCell.style.marginLeft = cellSpacing + "px";
                }
                rowElement.appendChild(renderedCell);
            }
        }
        return rowElement.children.length > 0 ? rowElement : undefined;
    }
    shouldSerialize(_context) {
        return true;
    }
    addCell(cell) {
        this.internalAddItem(cell);
    }
    removeCellAt(columnIndex) {
        if (columnIndex >= 0 && columnIndex < this.getItemCount()) {
            return this.removeItem(this.getItemAt(columnIndex));
        }
        return false;
    }
    indexOf(cardElement) {
        return cardElement instanceof TableCell ? this.internalIndexOf(cardElement) : -1;
    }
    ensureHasEnoughCells(cellCount) {
        while (this.getItemCount() < cellCount) {
            this.addCell(new TableCell());
        }
    }
    getJsonTypeName() {
        return "TableRow";
    }
    getIsFirstRow() {
        return this.parentTable.getItemAt(0) === this;
    }
    get parentTable() {
        return this.parent;
    }
    get isStandalone() {
        return false;
    }
}
//#region Schema
TableRow.styleProperty = new ContainerStyleProperty(Versions.v1_5, "style");
TableRow.horizontalCellContentAlignmentProperty = new EnumProperty(Versions.v1_5, "horizontalCellContentAlignment", HorizontalAlignment);
TableRow.verticalCellContentAlignmentProperty = new EnumProperty(Versions.v1_5, "verticalCellContentAlignment", VerticalAlignment);
__decorate([
    property(TableRow.horizontalCellContentAlignmentProperty)
], TableRow.prototype, "horizontalCellContentAlignment", void 0);
__decorate([
    property(TableRow.verticalCellContentAlignmentProperty)
], TableRow.prototype, "verticalCellContentAlignment", void 0);
export class Table extends StylableContainer {
    constructor() {
        //#region Schema
        super(...arguments);
        this._columns = [];
        this.firstRowAsHeaders = true;
        this.showGridLines = true;
    }
    get gridStyle() {
        const style = this.getValue(Table.gridStyleProperty);
        if (style && this.hostConfig.containerStyles.getStyleByName(style)) {
            return style;
        }
        return undefined;
    }
    set gridStyle(value) {
        this.setValue(Table.gridStyleProperty, value);
    }
    //#endregion
    ensureRowsHaveEnoughCells() {
        for (let i = 0; i < this.getItemCount(); i++) {
            this.getItemAt(i).ensureHasEnoughCells(this.getColumnCount());
        }
    }
    removeCellsFromColumn(columnIndex) {
        for (let i = 0; i < this.getItemCount(); i++) {
            this.getItemAt(i).removeCellAt(columnIndex);
        }
    }
    getCollectionPropertyName() {
        return "rows";
    }
    createItemInstance(typeName) {
        return !typeName || typeName === "TableRow" ? new TableRow() : undefined;
    }
    internalParse(source, context) {
        super.internalParse(source, context);
        this.ensureRowsHaveEnoughCells();
    }
    internalRender() {
        if (this.getItemCount() > 0) {
            let totalWeights = 0;
            for (const column of this._columns) {
                if (column.width.unit === SizeUnit.Weight) {
                    totalWeights += column.width.physicalSize;
                }
            }
            for (const column of this._columns) {
                if (column.width.unit === SizeUnit.Pixel) {
                    column.computedWidth = new SizeAndUnit(column.width.physicalSize, SizeUnit.Pixel);
                }
                else {
                    column.computedWidth = new SizeAndUnit((100 / totalWeights) * column.width.physicalSize, SizeUnit.Weight);
                }
            }
            const tableElement = document.createElement("div");
            tableElement.setAttribute("role", "table");
            tableElement.style.display = "flex";
            tableElement.style.flexDirection = "column";
            if (this.showGridLines) {
                const styleDefinition = this.hostConfig.containerStyles.getStyleByName(this.gridStyle);
                if (styleDefinition.borderColor) {
                    const borderColor = stringToCssColor(styleDefinition.borderColor);
                    if (borderColor) {
                        tableElement.style.borderTop = "1px solid " + borderColor;
                        tableElement.style.borderLeft = "1px solid " + borderColor;
                    }
                }
            }
            const cellSpacing = this.hostConfig.table.cellSpacing;
            for (let i = 0; i < this.getItemCount(); i++) {
                const renderedRow = this.getItemAt(i).render();
                if (renderedRow) {
                    if (i > 0 && !this.showGridLines && cellSpacing > 0) {
                        const separatorRow = document.createElement("div");
                        separatorRow.setAttribute("aria-hidden", "true");
                        separatorRow.style.height = cellSpacing + "px";
                        tableElement.appendChild(separatorRow);
                    }
                    tableElement.appendChild(renderedRow);
                }
            }
            return tableElement;
        }
        return undefined;
    }
    addColumn(column) {
        this._columns.push(column);
        this.ensureRowsHaveEnoughCells();
    }
    removeColumn(column) {
        const index = this._columns.indexOf(column);
        if (index >= 0) {
            this.removeCellsFromColumn(index);
            this._columns.splice(index, 1);
        }
    }
    getColumnCount() {
        return this._columns.length;
    }
    getColumnAt(index) {
        return this._columns[index];
    }
    addRow(row) {
        this.internalAddItem(row);
        row.ensureHasEnoughCells(this.getColumnCount());
    }
    indexOf(cardElement) {
        return cardElement instanceof TableRow ? this.internalIndexOf(cardElement) : -1;
    }
    getJsonTypeName() {
        return "Table";
    }
}
Table._columnsProperty = new SerializableObjectCollectionProperty(Versions.v1_5, "columns", TableColumnDefinition);
Table.firstRowAsHeadersProperty = new BoolProperty(Versions.v1_5, "firstRowAsHeaders", true);
Table.showGridLinesProperty = new BoolProperty(Versions.v1_5, "showGridLines", true);
Table.gridStyleProperty = new ContainerStyleProperty(Versions.v1_5, "gridStyle");
Table.horizontalCellContentAlignmentProperty = new EnumProperty(Versions.v1_5, "horizontalCellContentAlignment", HorizontalAlignment);
Table.verticalCellContentAlignmentProperty = new EnumProperty(Versions.v1_5, "verticalCellContentAlignment", VerticalAlignment);
__decorate([
    property(Table._columnsProperty)
], Table.prototype, "_columns", void 0);
__decorate([
    property(Table.firstRowAsHeadersProperty)
], Table.prototype, "firstRowAsHeaders", void 0);
__decorate([
    property(Table.showGridLinesProperty)
], Table.prototype, "showGridLines", void 0);
__decorate([
    property(Table.gridStyleProperty)
], Table.prototype, "gridStyle", null);
__decorate([
    property(Table.horizontalCellContentAlignmentProperty)
], Table.prototype, "horizontalCellContentAlignment", void 0);
__decorate([
    property(Table.verticalCellContentAlignmentProperty)
], Table.prototype, "verticalCellContentAlignment", void 0);
GlobalRegistry.defaultElements.register("Table", Table, Versions.v1_5);
