// @pnp/cli-microsoft365/dist/m365/spo/commands/list/ListTemplateType.js
export var ListTemplateType;
(function (ListTemplateType) {
    ListTemplateType[ListTemplateType["GenericList"] = 100] = "GenericList";
    ListTemplateType[ListTemplateType["DocumentLibrary"] = 101] = "DocumentLibrary";
    ListTemplateType[ListTemplateType["Survey"] = 102] = "Survey";
    ListTemplateType[ListTemplateType["Links"] = 103] = "Links";
    ListTemplateType[ListTemplateType["Announcements"] = 104] = "Announcements";
    ListTemplateType[ListTemplateType["Contacts"] = 105] = "Contacts";
    ListTemplateType[ListTemplateType["Events"] = 106] = "Events";
    ListTemplateType[ListTemplateType["Tasks"] = 107] = "Tasks";
    ListTemplateType[ListTemplateType["DiscussionBoard"] = 108] = "DiscussionBoard";
    ListTemplateType[ListTemplateType["PictureLibrary"] = 109] = "PictureLibrary";
    ListTemplateType[ListTemplateType["DataSources"] = 110] = "DataSources";
    ListTemplateType[ListTemplateType["XmlForm"] = 115] = "XmlForm";
    ListTemplateType[ListTemplateType["NoCodeWorkflows"] = 117] = "NoCodeWorkflows";
    ListTemplateType[ListTemplateType["WorkflowProcess"] = 118] = "WorkflowProcess";
    ListTemplateType[ListTemplateType["WebPageLibrary"] = 119] = "WebPageLibrary";
    ListTemplateType[ListTemplateType["CustomGrid"] = 120] = "CustomGrid";
    ListTemplateType[ListTemplateType["WorkflowHistory"] = 140] = "WorkflowHistory";
    ListTemplateType[ListTemplateType["GanttTasks"] = 150] = "GanttTasks";
    ListTemplateType[ListTemplateType["IssuesTracking"] = 1100] = "IssuesTracking";
})(ListTemplateType || (ListTemplateType = {}));
