// @pnp/cli-microsoft365/dist/m365/spo/commands/brandcenter/brandcenter-colors-list.js
import commands from '../../commands.js';
import { globalOptionsZod } from '../../../../Command.js';
import request from '../../../../request.js';
import { odata } from '../../../../utils/odata.js';
import { spo } from '../../../../utils/spo.js';
import SpoCommand from '../../../base/SpoCommand.js';
const options = globalOptionsZod.strict();
class SpoBrandCenterColorsListCommand extends SpoCommand {
    get name() {
        return commands.BRANDCENTER_COLORS_LIST;
    }
    get description() {
        return 'Lists the brand center colors';
    }
    defaultProperties() {
        return ['ColorCode', 'Title', 'IsVisible'];
    }
    get schema() {
        return options;
    }
    async commandAction(logger) {
        if (this.verbose) {
            await logger.logToStderr(`Retrieving brand center colors...`);
        }
        try {
            const spoUrl = await spo.getSpoUrl(logger, this.verbose);
            const configRequestOptions = {
                url: `${spoUrl}/_api/Brandcenter/Configuration`,
                headers: {
                    accept: 'application/json;odata=nometadata'
                },
                responseType: 'json'
            };
            const config = await request.get(configRequestOptions);
            if (!config.BrandColorsListUrl) {
                if (this.verbose) {
                    await logger.logToStderr('Brand colors list not found.');
                }
                await logger.log([]);
                return;
            }
            if (this.verbose) {
                await logger.logToStderr(`Brand colors list found at '${config.BrandColorsListUrl.DecodedUrl}'.`);
            }
            const items = await odata.getAllItems(`${config.SiteUrl}/_api/web/lists(guid'${config.BrandColorsListId}')/items?$select=OData__SPColorTitle,OData__SPColorCode,OData__SPColorVisible`);
            const result = items.map(item => ({
                Title: item.OData__SPColorTitle,
                ColorCode: item.OData__SPColorCode,
                IsVisible: item.OData__SPColorVisible
            }));
            await logger.log(result);
        }
        catch (err) {
            this.handleRejectedODataJsonPromise(err);
        }
    }
}
export default new SpoBrandCenterColorsListCommand();
